Exceptions with Variable Messages Get Collapsed

<!--

Thanks for opening an issue! To help the Numba team handle your information
efficiently, please first ensure that there is no other issue present that
already describes the issue you have
(search at https://github.com/numba/numba/issues?&q=is%3Aissue).

-->

## Reporting a bug

<!--

Before submitting a bug report please ensure that you can check off these boxes:

-->
**Versions Tested:**
Tested on 0.66.0 and 0.65.1 in python version 3.14.6

**Brief Description:**
A function compiled with a jit decorator collapses all error messages stored in a variable with the same name and the same type of exception (instead of string literals passed directly to the exception) to all be exactly the same:

**Reproducing example:**
```
from numba import njit
import pytest

@njit()
def fail_message(a):
    if a == 0:
        msg = 'A was 0'
        raise ValueError(msg)
    if a == 1:
        msg = 'A was 1'
        raise ValueError(msg)
    if a==2:
        msg = 'A was 2'
        raise ValueError(msg)

@pytest.mark.parametrize('itr', [0,1,2])
def test_message(itr) -> None:
    with pytest.raises(ValueError, match=f'A was {itr}'):
        fail_message(itr)
```
**Expected behavior:** All tests pass
**Actual behavior:** 2/3 tests fail, all examples print 'A was 0':
```
numba_message_test.py .FF                                                                                                                                                                       [100%]

============================================================================================== FAILURES ===============================================================================================
___________________________________________________________________________________________ test_message[1] ___________________________________________________________________________________________

itr = 1

    @pytest.mark.parametrize('itr', [0,1,2])
    def test_message(itr) -> None:
>       with pytest.raises(ValueError, match=f'A was {itr}'):
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       AssertionError: Regex pattern did not match.
E         Expected regex: 'A was 1'
E         Actual message: 'A was 0'

numba_message_test.py:18: AssertionError
___________________________________________________________________________________________ test_message[2] ___________________________________________________________________________________________

itr = 2

    @pytest.mark.parametrize('itr', [0,1,2])
    def test_message(itr) -> None:
>       with pytest.raises(ValueError, match=f'A was {itr}'):
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       AssertionError: Regex pattern did not match.
E         Expected regex: 'A was 2'
E         Actual message: 'A was 0'

numba_message_test.py:18: AssertionError
======================================================================================= short test summary info =======================================================================================
FAILED numba_message_test.py::test_message[1] - AssertionError: Regex pattern did not match.
FAILED numba_message_test.py::test_message[2] - AssertionError: Regex pattern did not match.
```
**What this is an Issue:**
This way of writing exception messages is _required_ by ruff rule EM101, so numba-compiled code that complies with ruff rule EM101 that raises the same type of exception under different conditions will not produce meaningful error messages, and will have qualitatively different behavior (passing/failing unit tests) with and without the decorator.

--- a/docs/upcoming_changes/10702.bug_fix.rst
+++ b/docs/upcoming_changes/10702.bug_fix.rst
@@ -1,0 +1,5 @@
+Fix Dynamic Raise Exception Message Handling
+--------------------------------------------
+
+Raising a dynamic exception with a literal message stored in a variable could cause the compiled code to display a different literal value.
+Numba now handles dynamic exception message variables self-consistently, and prevents message collisions between variables with the same underlying storage representation.

--- a/numba/core/callconv.py
+++ b/numba/core/callconv.py
@@ -557,7 +557,7 @@ class CPUCallConv(BaseCallConv):
         #                          ^ Number of dynamic arguments = 2
         #
 
-        _hash = hashlib.sha1(str(st_type).encode()).hexdigest()
+        _hash = hashlib.sha1(str((st_type, nb_types)).encode()).hexdigest()
         name = f'__excinfo_unwrap_args{_hash}'
         if name in module.globals:
             return module.globals.get(name)

--- a/numba/core/ir_utils.py
+++ b/numba/core/ir_utils.py
@@ -304,6 +304,8 @@ def visit_vars_stmt(stmt, callback, cbdata):
         stmt.value = visit_vars_inner(stmt.value, callback, cbdata)
     elif isinstance(stmt, ir.Raise):
         stmt.exception = visit_vars_inner(stmt.exception, callback, cbdata)
+    elif isinstance(stmt, (ir.DynamicRaise, ir.DynamicTryRaise)):
+        stmt.exc_args = visit_vars_inner(stmt.exc_args, callback, cbdata)
     elif isinstance(stmt, ir.Branch):
         stmt.cond = visit_vars_inner(stmt.cond, callback, cbdata)
     elif isinstance(stmt, ir.Jump):

--- a/numba/tests/test_exceptions.py
+++ b/numba/tests/test_exceptions.py
@@ -434,6 +434,36 @@ class TestRaising(TestCase):
                     raise_literal_message(i)
                 self.assertEqual(str(raises.exception), f"A was {i}")
 
+    def test_dynamic_raise_single_literal_message(self):
+        @njit
+        def raise_single_literal_message(arg):
+            msg = "False"
+            if arg:
+                msg = "True"
+            raise ValueError(msg)
+
+        for arg in (False, True):
+            with self.subTest(arg=arg):
+                with self.assertRaises(ValueError) as raises:
+                    raise_single_literal_message(arg)
+                self.assertEqual(str(raises.exception), str(arg))
+
+    def test_dynamic_raise_mixed_int_width(self):
+        @njit
+        def raise_mixed_int_width(a):
+            if a == 0:
+                value = np.int64(-7)
+                raise ValueError(value)
+            value = np.uint64(2 ** 63)
+            raise ValueError(value)
+
+        cases = [(0, "-7"), (1, str(2 ** 63))]
+        for arg, expected in cases:
+            with self.subTest(arg=arg):
+                with self.assertRaises(ValueError) as raises:
+                    raise_mixed_int_width(arg)
+                self.assertEqual(str(raises.exception), expected)
+
     def test_disable_nrt(self):
         @njit(_nrt=False)
         def raise_with_no_nrt(i):

--- a/numba/tests/test_exceptions.py
+++ b/numba/tests/test_exceptions.py
@@ -415,6 +415,25 @@ class TestRaising(TestCase):
         with self.assertRaisesRegex(ValueError, msg):
             raise_literal_dict2()
 
+    def test_dynamic_raise_literal_message(self):
+        @njit
+        def raise_literal_message(a):
+            if a == 0:
+                msg = "A was 0"
+                raise ValueError(msg)
+            if a == 1:
+                msg = "A was 1"
+                raise ValueError(msg)
+            if a == 2:
+                msg = "A was 2"
+                raise ValueError(msg)
+
+        for i in range(3):
+            with self.subTest(i=i):
+                with self.assertRaises(ValueError) as raises:
+                    raise_literal_message(i)
+                self.assertEqual(str(raises.exception), f"A was {i}")
+
     def test_disable_nrt(self):
         @njit(_nrt=False)
         def raise_with_no_nrt(i):

